frontend/src/metabase/lib/data_grid.js

```diff
@@ -104,76 +76,118 @@ export function multiLevelPivot(
-    if (valueColumns.length > 1) {
-      const colNames = valueColumns.map(col => ({
-        value: col.display_name,
-        span: 1,
-      }));
-      topIndex.push([rowTotals, colNames]);
-    } else {
-      topIndex.push([rowTotals]);
-    }
+    formattedColumnTreeWithoutValues.push({
+      value: t`Row totals`,
+      children: [],
+      isSubtotal: true,
+      isGrandTotal: true,
+    });
   }
 
-  const formattedRowTree = formatValuesInTree(
+  const columnIndex = addEmptyIndexItem(
+    formattedColumnTreeWithoutValues.flatMap(enumeratePaths),
+  );
+  const formattedColumnTree = addValueColumnNodes(
+    formattedColumnTreeWithoutValues,
+    valueColumns,
+  );
+
+  const formattedRowTreeWithoutSubtotals = formatValuesInTree(
     rowColumnTree,
     leftIndexFormatters,
   );
-  const leftIndex = addSubtotals(formattedRowTree, leftIndexFormatters);
-  if (leftIndex.length > 1) {
-    // if there are multiple rows, we should add another for grand totals
-    leftIndex.push([
-      {
-        value: t`Grand totals`,
-        span: 1,
-        isSubtotal: true,
-        isGrandTotal: true,
-        children: [],
-      },
-    ]);
+  const formattedRowTree = addSubtotals(
+    formattedRowTreeWithoutSubtotals,
+    leftIndexFormatters,
+  );
+  if (formattedRowTreeWithoutSubtotals.length > 1) {
+    // if there are multiple columns, we should add another for row totals
+    formattedRowTree.push({
+      value: t`Grand totals`,
+      isSubtotal: true,
+      isGrandTotal: true,
+      children: [],
+    });
   }
 
-  // we need at least one row/column, so convert zero length to 1
-  const columnCount = topIndex.length || 1;
-  const rowCount = leftIndex.length || 1;
+  const rowIndex = addEmptyIndexItem(formattedRowTree.flatMap(enumeratePaths));
+
+  const leftHeaderItems = treeToArray(formattedRowTree.flat());
+  const topHeaderItems = treeToArray(formattedColumnTree.flat());
+
+  const getRowSection = createRowSectionGetter({
+    valuesByKey,
+    subtotalValues,
+    valueFormatters,
+    columnColumnIndexes,
+    rowColumnIndexes,
+    columnIndex,
+    rowIndex,
+  });
+
   return {
-    topIndex,
-    leftIndex,
-    topIndexFormatters,
-    rowColumnTree,
-    leftIndexFormatters,
-    columnCount,
-    rowCount,
-    getRowSection: createRowSectionGetter({
-      valuesByKey,
-      columnColumnTree,
-      rowColumnTree,
-      valueFormatters,
-      subtotalValues,
-      columnColumnIndexes,
-      rowColumnIndexes,
-    }),
+    leftHeaderItems,
+    topHeaderItems,
+    rowCount: rowIndex.length,
+    columnCount: columnIndex.length,
+    getRowSection,
   };
 }
 
+// This pulls apart the different aggregations that were packed into one result set.
+// There's a column indicating which breakouts were used to compute that row.
+// We use that column to split apart the data and convert the field refs to indexes.
+function splitPivotData(data, rowIndexes, columnIndexes) {
```

Comment:
Looks like this function isn't modified, just moved, right?
Any reason to do so? Seems that it's natural for `splitPivotData` to appear first before being used in `multiLevelPivot`.

Reply:
Yeah that's right it wasn't changed. I moved it since all the other functions called from `multiLevelPivot` follow it.

I go back and forth, but my general preference is to have the high level function proceed their building blocks. I don't have strong feelings though.

Reply:
Roger. I was just confused by the diff view.